internal/controller/acl/acl.go

```diff
@@ -0,0 +1,207 @@
+		WithOptions(o).
+		For(&v1alpha1.AccessControlList{}).
+		Complete(r)
+}
+
+// A connector is expected to produce an ExternalClient when its Connect method
+// is called.
+type connector struct {
+	kube         client.Client
+	usage        resource.Tracker
+	log          logging.Logger
+	newServiceFn func(creds []byte) (*kadm.Client, error)
+}
+
+// Connect typically produces an ExternalClient by:
+// 1. Tracking that the managed resource is using a ProviderConfig.
+// 2. Getting the managed resource's ProviderConfig.
+// 3. Getting the credentials specified by the ProviderConfig.
+// 4. Using the credentials to form a client.
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.AccessControlList)
+	if !ok {
+		return nil, errors.New(errNotAccessControlList)
+	}
+
+	if err := c.usage.Track(ctx, mg); err != nil {
+		return nil, errors.Wrap(err, errTrackPCUsage)
+	}
+
+	pc := &apisv1alpha1.ProviderConfig{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Name: cr.GetProviderConfigReference().Name}, pc); err != nil {
+		return nil, errors.Wrap(err, errGetPC)
+	}
+
+	cd := pc.Spec.Credentials
+	data, err := resource.CommonCredentialExtractor(ctx, cd.Source, c.kube, cd.CommonCredentialSelectors)
+	if err != nil {
+		return nil, errors.Wrap(err, errGetCreds)
+	}
+
+	svc, err := c.newServiceFn(data)
+	if err != nil {
+		return nil, errors.Wrap(err, errNewClient)
+	}
+
+	return &external{kafkaClient: svc, log: c.log}, nil
+}
+
+// An ExternalClient observes, then either creates, updates, or deletes an
+// external resource to ensure it reflects the managed resource's desired state.
+type external struct {
+	kafkaClient *kadm.Client
+	log         logging.Logger
+}
+
+func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.AccessControlList)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotAccessControlList)
+	}
+
+	// Check if the external name is set, to determine if ACL has been created or not
+	ext := meta.GetExternalName(cr)
+	if ext == "" {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	extname, err := acl.ConvertFromJSON(meta.GetExternalName(cr))
+	compare := acl.CompareAcls(*extname, *acl.Generate(cr.Name, &cr.Spec.ForProvider))
+
+	if !compare {
+		return managed.ExternalObservation{
+			ResourceExists:   true,
+			ResourceUpToDate: false,
+		}, errors.Wrap(err, "Updating not allowed")
+	}
+
+	ae, err := acl.List(ctx, c.kafkaClient, extname)
+
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errListACL)
+	}
+
+	if ae == nil {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	cr.Status.AtProvider.ID = ae.Name
+	cr.Status.SetConditions(v1.Available())
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        acl.IsUpToDate(&cr.Spec.ForProvider, ae),
```

Comment:
I think we no longer need to check if ACL is up to date at this point. We don't support updating and already errored above.

Reply:
```suggestion
		ResourceUpToDate:        true,
```